README.md
# Options Testing Framework

This library is a framework for testing stock options strategies. It works along with the popular Backtrader framework. The Backtrader framework is an excellent solution for backtesting stock trading ideas. 
However, testing options trading strategies is much more complicated. This
framework will work with backtrader to test your options trading strategies.

## Getting Started

- clone this repo
- ensure pipenv is installed
  - `pip install pipenv`
- run `pipenv install --dev` to install the dependencies
- run `pipenv shell` to activate the virtual environment
- run `pytest` to run the tests
  - `pytest -vv` to run the tests with more verbosity
- run `pcmd` to run the CLI

## CLI Examples

### Entry Point 

![entry point](/images/cli/entry_point.png)


### Backtesting Subcommand

sub command

![backtesting subcommand](/images/cli/sub_command_backtest.png)

Vertical Spread Example + Butterfly Strategy Example

![sub command example](/images/cli/sub_command_backtest_details.png)

### Config Subcommand

![config](/images/cli/sub_command_config_show.png)

changelog.md
# Change log

## 2023-11-09 - v0.1.0.c

- refactored 105 passing tests to use the new `conftest.py` fixtures file
- there are some tests that still need to be implemented to use the new fixtures


## 2023-11-09 - v0.1.0.b

### Added
- Added new test file `tests/test_enums.py` to verify the dynaconf implementation

### Modified
- Updated `settings.toml` now contains option type enums

## 2023-11-09 - v0.1.0.a

- Update README.md with pytest example
- Added `pytest.ini`
  - pointed default test dir to `tests` so the tests archive doesn't get run
  - added options by default that show STDOUT, increase test verbosity, and show short tracebacks 
- Added `tests/test_silly.py`
  - a silly test to make sure pytest is working
- Added `tests_archive/` directory
  - Archived `tests/options_test_helper.py`
  - Archived `tests/test_data/L2_options_20230301.csv` and similar files
  - Archived `tests/test_data/spx_11_01_2021.csv` and similar files
  - Archived `tests/test_option.py`
  - Archived `tests/test_option_chain.py`
  - Archived `tests/test_option_combo.py`
  - Archived `tests/test_single_option.py`
  - Archived `tests/test_test.py`
  - Archived `tests/test_vertical_spread.py`

## 2023-11-09 - v0.1.0

- update README.md with CLI examples
- create change log file
- added pipenv support
- added black formatter
- added pylint
- added dynaconf to simplify configuration management
- added a basic CLI
- created a `src` directory and a `setup.py` so the project is installable using `pip install -e .`
- modified .gitignore to include dynaconf secrets file
- created a `template.secrets.toml` file for to document dynaconf secrets

tests/conftest.py
from datetime import datetime
from collections import namedtuple
import pytest
from options_framework.config import settings
from options_framework.option import Option

Quote = namedtuple("Quote", "quote_date spot_price bid ask price")


@pytest.fixture
def expiration():
    """Fixture for expiration date"""
    return datetime.strptime("07-16-2021", "%m-%d-%Y")


@pytest.fixture
def quote_date():
    """Fixture for quote date"""
    return datetime.strptime("2021-07-01 09:45:00.000000", "%Y-%m-%d %H:%M:%S.%f")


@pytest.fixture
def update_quote_date():
    """Fixture for update quote date"""
    return datetime.strptime("2021-07-02 09:45:00.000000", "%Y-%m-%d %H:%M:%S.%f")


@pytest.fixture
def update_quote_date2():
    """Fixture for second update quote date"""
    return datetime.strptime("2021-07-02 14:31:00.000000", "%Y-%m-%d %H:%M:%S.%f")


@pytest.fixture
def update_quote_date3():
    """Fixture for third update quote date"""
    return datetime.strptime("2021-07-16 11:31:00.000000", "%Y-%m-%d %H:%M:%S.%f")


@pytest.fixture
def at_expiration_quote_date():
    """Fixture for at expiration quote date"""
    return datetime.strptime("2021-07-16 16:15:00.000000", "%Y-%m-%d %H:%M:%S.%f")


@pytest.fixture
def standard_fee():
    """Fixture for standard fee"""
    return 0.50


@pytest.fixture
def ticker():
    """Fixture for ticker"""
    return "XYZ"


@pytest.fixture
def call_option(quote_date, ticker, expiration):
    """Fixture for call option"""
    _id, strike, spot_price, bid, ask, price = (1, 100.0, 90.0, 1.0, 2.0, 1.5)
    return Option(
        _id,
        ticker,
        strike,
        expiration,
        settings.OptionType.CALL,
        quote_date=quote_date,
        spot_price=spot_price,
        bid=bid,
        ask=ask,
        price=price,
    )


@pytest.fixture
def call_option_update_values_1(update_quote_date):
    """Fixture for first set of call option update values"""
    quote = Quote(
        quote_date=update_quote_date,
        spot_price=110.0,
        bid=9.50,
        ask=10.5,
        price=10.00,
    )  # ITM
    return quote.quote_date, quote.spot_price, quote.bid, quote.ask, quote.price


@pytest.fixture
def call_option_update_values_2(update_quote_date2):
    """Fixture for second set of call option update values"""
    quote = Quote(
        quote_date=update_quote_date2,
        spot_price=105.0,
        bid=4.50,
        ask=5.5,
        price=5.00,
    )  # ITM
    return quote.quote_date, quote.spot_price, quote.bid, quote.ask, quote.price


@pytest.fixture
def call_option_update_values_3(update_quote_date3):
    """Fixture for third set of call option update values"""
    quote = Quote(
        quote_date=update_quote_date3,
        spot_price=90.0,
        bid=0,
        ask=0.05,
        price=0.03,
    )  # OTM
    return quote.quote_date, quote.spot_price, quote.bid, quote.ask, quote.price


@pytest.fixture
def call_option_extended_properties(ticker, expiration, quote_date):
    """Fixture for call option with extended properties"""
    _id = 1
    strike = 100
    spot_price, bid, ask, price = (110.0, 1.0, 2.0, 1.5)
    delta, gamma, theta, vega, open_interest, rho, iv = (
        0.3459,
        -0.1234,
        0.0485,
        0.0935,
        100,
        0.132,
        0.3301,
    )
    return Option(
        _id,
        ticker,
        strike,
        expiration,
        settings.OptionType.CALL,
        quote_date=quote_date,
        spot_price=spot_price,
        bid=bid,
        ask=ask,
        price=price,
        delta=delta,
        gamma=gamma,
        theta=theta,
        vega=vega,
        rho=rho,
        implied_volatility=iv,
        open_interest=open_interest,
    )


@pytest.fixture
def put_option(ticker, expiration, quote_date):
    """Fixture for put option"""
    _id, strike, spot_price, bid, ask, price = (1, 100.0, 105.0, 1.0, 2.0, 1.5)
    return Option(
        _id,
        ticker,
        strike,
        expiration,
        settings.OptionType.PUT,
        quote_date=quote_date,
        spot_price=spot_price,
        bid=bid,
        ask=ask,
        price=price,
    )


@pytest.fixture
def put_option_update_values_1(update_quote_date):
    """Fixture for first set of put option update values"""
    quote = Quote(
        quote_date=update_quote_date,
        spot_price=90.0,
        bid=9.50,
        ask=10.5,
        price=10.00,
    )  # ITM
    return quote.quote_date, quote.spot_price, quote.bid, quote.ask, quote.price


@pytest.fixture
def put_option_update_values_2(update_quote_date2):
    """Fixture for second set of put option update values"""
    quote = Quote(
        quote_date=update_quote_date2,
        spot_price=95.0,
        bid=4.50,
        ask=5.5,
        price=5.00,
    )  # ITM
    return quote.quote_date, quote.spot_price, quote.bid, quote.ask, quote.price


@pytest.fixture
def put_option_update_values_3(update_quote_date3):
    """Fixture for third set of put option update values"""
    quote = Quote(
        quote_date=update_quote_date3,
        spot_price=110.0,
        bid=0,
        ask=0.05,
        price=0.03,
    )  # OTM
    return quote.quote_date, quote.spot_price, quote.bid, quote.ask, quote.price


@pytest.fixture
def put_option_with_extended_properties(ticker, expiration, quote_date):
    """Fixture for put option with extended properties"""
    _id = 1
    strike = 100
    spot_price, bid, ask, price = (105, 1.0, 2.0, 1.5)
    delta, gamma, theta, vega, open_interest, rho, iv = (
        -0.4492,
        -0.1045,
        0.0412,
        0.1143,
        900,
        0.0282,
        0.3347,
    )
    test_option = Option(
        _id,
        ticker,
        strike,
        expiration,
        settings.OptionType.PUT,
        quote_date=quote_date,
        spot_price=spot_price,
        bid=bid,
        ask=ask,
        price=price,
        delta=delta,
        gamma=gamma,
        theta=theta,
        vega=vega,
        rho=rho,
        implied_volatility=iv,
        open_interest=open_interest,
    )
    return test_option


tests/test_enums.py
"""
This module contains tests for the enums defined in the options_framework.config.settings module.
These tests ensure that the enum values do not change unexpectedly, which could cause issues in other parts of the codebase.
"""

from options_framework.config import settings


def test_the_enums():
    assert settings.OptionType.CALL == 1
    assert settings.OptionType.PUT == 2

    assert settings.TransactionType.BUY == 1
    assert settings.TransactionType.SELL == 2

    assert settings.OptionPositionType.SHORT == 1

    assert settings.OptionCombinationType.SINGLE == 1
    assert settings.OptionCombinationType.VERTICAL == 2
    assert settings.OptionCombinationType.RATIO == 3
    assert settings.OptionCombinationType.CALENDAR == 4
    assert settings.OptionCombinationType.DIAGONAL == 5
    assert settings.OptionCombinationType.STRADDLE == 6
    assert settings.OptionCombinationType.STRANGLE == 7
    assert settings.OptionCombinationType.BUTTERFLY == 8
    assert settings.OptionCombinationType.IRON_BUTTERFLY == 9
    assert settings.OptionCombinationType.CONDOR == 10
    assert settings.OptionCombinationType.IRON_CONDOR == 11
    assert settings.OptionCombinationType.COLLAR == 12
    assert settings.OptionCombinationType.CUSTOM == 100

    assert settings.OptionTradeType.CREDIT == 1
    assert settings.OptionTradeType.DEBIT == 2


tests/test_silly.py
def test_silliest():
    print()
    print("you should see this comment if pytest command was passed the -s flag")
    assert True


src/options_framework/utils/helpers.py
from decimal import Context, Decimal, getcontext


def decimalize_0(value):
    """
    Floating numbers can produce unexpected results in calculation.
    Decimal numbers fix this. This returns a decimal number with 0 decimal places.
    :param value: the floating point number to decimalize
    :return: a decimal number rounded to 0 decimal places
    """
    dec_val = Decimal(value)
    quantize_val = dec_val.quantize(Decimal("1"))
    return quantize_val


def decimalize_2(value):
    """
    Floating numbers can produce unexpected results in calculation.
    Decimal numbers fix this. This returns a decimal number with 2 decimal places.
    :param value: the floating point number to decimalize
    :return: a decimal number rounded to 2 decimal places
    """
    dec_val = Decimal(value)
    quantize_val = dec_val.quantize(Decimal("1.00"))
    return quantize_val


def decimalize_4(value):
    """
    Floating numbers can produce unexpected results in calculation.
    Decimal numbers fix this. This returns a decimal number with 4 decimal places.
    :param value: the floating point number to decimalize
    :return: a decimal number rounded to 4 decimal places
    """
    dec_val = Decimal(value)
    quantize_val = dec_val.quantize(Decimal("1.0000"))
    return quantize_val


def distinct(iterable):
    """
    Returns a list of distinct items from a given iterable
    :param iterable:
    :return: list of distinct items
    """
    distinct_values = set()
    for x in iterable:
        if x in distinct_values:
            continue
        yield x
        distinct_values.add(x)


options_framework_archive/spreads/single.py
from options_framework_archive.option_types import OptionType, OptionCombinationType, OptionPositionType, OptionTradeType
from options_framework_archive.spreads.option_combo import OptionCombination
from options_framework_archive.utils.helpers import decimalize_2

class Single(OptionCombination):

    def __init__(self, option, *args, **kwargs):
        self._option = option
        self._option_combination_type = OptionCombinationType.SINGLE
        super().__init__([option], args, kwargs)

    @property
    def option_trade_type(self):
        if self._option.position_type == OptionPositionType.LONG:
            return OptionTradeType.DEBIT
        else:
            return OptionTradeType.CREDIT

    def breakeven_price(self):
        trade_price = decimalize_2(self._option.trade_price)

        if self._option.option_type == OptionType.CALL:
            breakeven_price = self._option.strike + trade_price
        else:
            breakeven_price = self._option.strike - trade_price

        return float(breakeven_price)



options_framework_archive/spreads/vertical.py
from options_framework_archive.option_types import OptionPositionType, OptionType, OptionTradeType, OptionCombinationType
from options_framework_archive.spreads.option_combo import OptionCombination
from options_framework_archive.utils.helpers import decimalize_0, decimalize_2


class Vertical(OptionCombination):

    def __init__(self, long_option, short_option, *args, **kwargs):
        if not long_option.position_type == OptionPositionType.LONG:
            raise ValueError("The long_option parameter must be a long option.")
        if not short_option.position_type == OptionPositionType.SHORT:
            raise ValueError("The short_option parameter must be a short option.")
        if long_option.option_type != short_option.option_type:
            raise ValueError("A vertical spread must contain options of the same type. Both must be calls or puts.")
        if long_option.expiration != short_option.expiration:
            raise ValueError("All options in the spread must have the same expiration")
        if abs(long_option.quantity) != abs(short_option.quantity):
            raise ValueError("Both options must have the same absolute quantity for both long and short legs")
        if long_option.strike == short_option.strike:
            raise ValueError("The options provided do not form a vertical spread")

        self._option_combination_type = OptionCombinationType.VERTICAL
        self._option_type = long_option.option_type

        # determine if this is a debit or credit spread
        if self._option_type == OptionType.CALL:
            self._option_trade_type = OptionTradeType.DEBIT if long_option.strike < short_option.strike \
                else OptionTradeType.CREDIT
        elif self._option_type == OptionType.PUT:
            self._option_trade_type = OptionTradeType.DEBIT if long_option.strike > short_option.strike \
                else OptionTradeType.CREDIT

        super().__init__([long_option, short_option], args, kwargs)
        self._long_option = self._options[0]
        self._short_option = self._options[1]

    @property
    def option_trade_type(self):
        return self._option_trade_type

    @property
    def option_type(self):
        return self._option_type

    def max_profit(self):
        # call debit: Max profit = the spread between the strike prices - net premium paid.
        # call credit: Max profit = net premium received.
        # put debit: Max profit = the spread between the strike prices - net premium paid.
        # put credit: Max profit = net premium received.
        long_price = decimalize_2(self._long_option.trade_price)
        short_price = decimalize_2(self._short_option.trade_price)
        if self._option_trade_type == OptionTradeType.DEBIT:
            max_profit = ((self.spread_width() - (long_price-short_price)) * 100
                                * self._long_option.quantity)
        else:
            max_profit = (long_price-short_price)*100*self._short_option.quantity
        return float(max_profit)

    def max_loss(self):
        long_price = decimalize_2(self._long_option.trade_price)
        short_price = decimalize_2(self._short_option.trade_price)
        quantity = decimalize_0(self._long_option.quantity)
        if self._option_trade_type == OptionTradeType.DEBIT:
            max_loss = (long_price-short_price) * 100 * quantity
        else:
            max_loss = ((self.spread_width() - (short_price - long_price)) * 100 * quantity)

        return float(max_loss)

    def spread_width(self):
        return abs(self._long_option.strike - self._short_option.strike)

    def total_premium(self):
        # long_price = decimalize_2(self._long_option.trade_price)
        # short_price = decimalize_2(self._short_option.trade_price)
        # quantity = decimalize_0(self._long_option.quantity)
        # premium = (long_price-short_price) * 100 * quantity
        total_premium = self.premium() * 100 * self._long_option.quantity
        return total_premium

    def breakeven_price(self):
        long_price = decimalize_2(self._long_option.trade_price)
        short_price = decimalize_2(self._short_option.trade_price)
        breakeven_price = None
        # Call debit spread: Breakeven point = long call's strike price + net premium paid.
        if self._option_trade_type == OptionTradeType.DEBIT and self._option_type == OptionType.CALL:
            breakeven_price = self._long_option.strike + (long_price - short_price)
        # Call credit spread: Breakeven point = short call's strike price + net premium received.
        elif self._option_trade_type == OptionTradeType.CREDIT and self._option_type == OptionType.CALL:
            breakeven_price = self._short_option.strike + (short_price - long_price)
        # Put debit spread: Breakeven point = long put's strike price - net premium paid.
        elif self._option_trade_type == OptionTradeType.DEBIT and self._option_type == OptionType.PUT:
            breakeven_price = self._long_option.strike - (long_price - short_price)
        # Put credit spread: Breakeven point = short put's strike price - net premium received.
        elif self._option_trade_type == OptionTradeType.CREDIT and self._option_type == OptionType.PUT:
            breakeven_price = self._short_option.strike - (short_price - long_price)

        return float(breakeven_price)


tests_archive/options_test_helper.py
import datetime
from options_framework_archive.option_types import OptionType
from options_framework_archive.option import Option

test_expiration = datetime.datetime.strptime("07-16-2021", "%m-%d-%Y")
test_quote_date = datetime.datetime.strptime("2021-07-01 09:45:00.000000", "%Y-%m-%d %H:%M:%S.%f")
test_update_quote_date = datetime.datetime.strptime("2021-07-02 09:45:00.000000", "%Y-%m-%d %H:%M:%S.%f")
test_update_quote_date2 = datetime.datetime.strptime("2021-07-02 14:31:00.000000", "%Y-%m-%d %H:%M:%S.%f")
test_update_quote_date3 = datetime.datetime.strptime("2021-07-16 11:31:00.000000", "%Y-%m-%d %H:%M:%S.%f")
at_expiration_quote_date = datetime.datetime.strptime("2021-07-16 16:15:00.000000", "%Y-%m-%d %H:%M:%S.%f")
standard_fee = 0.50
ticker = "XYZ"

def get_test_call_option():
    _id, strike, spot_price, bid, ask, price = (1, 100.0, 90.0, 1.0, 2.0, 1.5)
    test_option = Option(id, ticker, strike, test_expiration, OptionType.CALL, quote_date=test_quote_date,
                         spot_price=spot_price, bid=bid, ask=ask, price=price)
    return test_option

def get_test_call_option_update_values_1():
    quote_date, spot_price, bid, ask, price = (test_update_quote_date, 110.0, 9.50, 10.5, 10.00)  # ITM
    return quote_date, spot_price, bid, ask, price

def get_test_call_option_update_values_2():
    quote_date, spot_price, bid, ask, price = (test_update_quote_date2, 105.0, 4.50, 5.5, 5.00)  # ITM
    return quote_date, spot_price, bid, ask, price

def get_test_call_option_update_values_3():
    quote_date, spot_price, bid, ask, price = (test_update_quote_date3, 90.0, 0, .05, 0.03)  # OTM
    return quote_date, spot_price, bid, ask, price

def get_test_call_option_extended_properties():
    _id = 1
    strike = 100
    spot_price, bid, ask, price = (110.0, 1.0, 2.0, 1.5)
    delta, gamma, theta, vega, open_interest, rho, iv = (0.3459, -0.1234, 0.0485, 0.0935, 100, 0.132, 0.3301)
    test_option = Option(id, ticker, strike, test_expiration, OptionType.CALL, quote_date=test_quote_date,
                         spot_price=spot_price, bid=bid, ask=ask, price=price,
                         delta=delta, gamma=gamma, theta=theta, vega=vega, rho=rho, implied_volatility=iv,
                         open_interest=open_interest)
    return test_option

def get_test_put_option():
    _id, strike, spot_price, bid, ask, price = (1, 100.0, 105.0, 1.0, 2.0, 1.5)
    test_option = Option(id, ticker, strike, test_expiration, OptionType.PUT, quote_date=test_quote_date,
                         spot_price=spot_price, bid=bid, ask=ask, price=price)
    return test_option

def get_test_put_option_update_values_1():
    quote_date, spot_price, bid, ask, price = (test_update_quote_date, 90.0, 9.50, 10.5, 10.00)  # ITM
    return quote_date, spot_price, bid, ask, price

def get_test_put_option_update_values_2():
    quote_date, spot_price, bid, ask, price = (test_update_quote_date2, 95.0, 4.50, 5.5, 5.00)  # ITM
    return quote_date, spot_price, bid, ask, price

def get_test_put_option_update_values_3():
    quote_date, spot_price, bid, ask, price = (test_update_quote_date3, 110.0, 0, .05, 0.03)  # OTM
    return quote_date, spot_price, bid, ask, price

def get_test_put_option_with_extended_properties():
    _id = 1
    strike = 100
    spot_price, bid, ask, price = (105, 1.0, 2.0, 1.5)
    delta, gamma, theta, vega, open_interest, rho, iv = (-0.4492, -0.1045, 0.0412, 0.1143, 900, 0.0282, 0.3347)
    test_option = Option(id, ticker, strike, test_expiration, OptionType.PUT, quote_date=test_quote_date,
                         spot_price=spot_price, bid=bid, ask=ask, price=price,
                         delta=delta, gamma=gamma, theta=theta, vega=vega, rho=rho, implied_volatility=iv,
                         open_interest=open_interest)
    return test_option
